content/articles/2017-11-15-withr-2.1.0.Rmd

```diff
@@ -0,0 +1,234 @@
+---
+title: withr 2.1.0
+date: '2017-11-15'
+slug: withr-2.1.0
+author: Jim Hester
+categories: [package]
+description: >
+  withr 2.1.0 is now available on CRAN.
+photo:
+  url: https://unsplash.com/photos/V1YEEItoOTE
+  author: Suganth
+---
+
+```{r setup, include = FALSE}
+knitr::opts_chunk$set(
+  comment = "#>",
+  collapse = TRUE
+)
+library(withr)
+```
+
+[withr 2.1.0](http://withr.r-lib.org) is now available on CRAN!
+[withr](http://withr.r-lib.org) makes working with global state in R safer and
+less error prone. It also has no non-base dependencies so is easy to include in
+packages.
+
+Install the latest version with:
+```{r, eval = FALSE}
+install.packages("withr")
+```
+
+# Why withr?
+
+Many functions in R modify global state in some fashion. Some common examples
+are `par()` for graphics parameters, `dir()` to change the current directory
+and `options()` to set a global option. Using these global options is useful
+when using R interactively, because you can set them early in your
+experimentation and they will remain set for the duration of the session.
+However this makes programming with these settings difficult, because in a
+pure function you should always reset the previous state when the function
+completes if you change a global setting.
+
+One common idiom for dealing with this problem is to save the current state,
+make your change, then reset the previous state.
+
+```{r}
+par("col" = "black")
+my_plot <- function(new) {
+  old <- par(col = "red", pch = 19)
+  plot(mtcars$hp, mtcars$wt)
+  par(old)
+}
+my_plot()
+par("col")
+```
+
+However one issue when there is an error before you are able to reset the
+options.
+
+```{r, error = TRUE}
+par("col" = "black")
+my_plot <- function(new) {
+  old <- par(col = "red", pch = 19)
+  plot(mtcars$hpp, mtcars$wt)
+  par(old)
+}
+my_plot()
+par("col")
+```
+
+Using the base function `on.exit()` is the robust solution to this problem,
+which will run the resetting code when the block is exited, regardless if it
+exits normally or with an error.
+
+```{r, error = TRUE}
+par("col" = "black")
+my_plot <- function(new) {
+  old <- par(col = "red", pch = 19)
+  on.exit(par(old))
+  plot(mtcars$hpp, mtcars$wt)
+}
+my_plot()
+par("col")
+```
+
+However this solution is somewhat cumbersome to work with. You
+need to remember to use an `on.exit()` call after each stateful call. In
+addition by default each `on.exit()` action will overwrite any previous
+`on.exit()` action in the same block (using `add = TRUE` prevents this). It is also
```

Comment:
In parenthetical, mention even with `on.exit()` the calls get cleaned up in the wrong order?